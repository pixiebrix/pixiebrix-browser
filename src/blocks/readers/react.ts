/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
