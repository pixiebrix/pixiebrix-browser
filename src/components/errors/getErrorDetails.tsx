/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { InputValidationError, OutputValidationError } from "@/blocks/errors";
import { getRootCause, isAxiosError } from "@/errors";
import { AxiosError } from "axios";
import { ErrorObject } from "serialize-error";
import InputValidationErrorDetail from "./InputValidationErrorDetail";
import NetworkErrorDetail from "./NetworkErrorDetail";
import OutputValidationErrorDetail from "./OutputValidationErrorDetail";
import { Col, Row } from "react-bootstrap";

type ErrorDetails = {
  title: string;
  detailsElement: React.ReactElement;
};

const getErrorDetails: (error: ErrorObject) => ErrorDetails = (error) => {
  const rootError = getRootCause(error);
  const { name, message } = rootError;

  let errorDetails: ErrorDetails;

  switch (name) {
    case "InputValidationError":
      errorDetails = {
        title: "Invalid inputs for block",
        detailsElement: (
          <InputValidationErrorDetail
            error={rootError as unknown as InputValidationError}
          />
        ),
      };
      break;
    case "OutputValidationError":
      errorDetails = {
        title: "Invalid output for block",
        detailsElement: (
          <OutputValidationErrorDetail
            error={rootError as unknown as OutputValidationError}
          />
        ),
      };
      break;
    case "ClientNetworkError": {
      const networkError: AxiosError = isAxiosError(rootError.error)
        ? rootError.error
        : (rootError as unknown as AxiosError);
      errorDetails = {
        title: "Network error",
        detailsElement: <NetworkErrorDetail error={networkError} />,
      };
      break;
    }

    default:
      errorDetails = {
        title: "Error",
        detailsElement: (
          <Row>
            <Col>
              {name}: {message}
            </Col>
          </Row>
        ),
      };
  }

  return errorDetails;
};

export default getErrorDetails;
