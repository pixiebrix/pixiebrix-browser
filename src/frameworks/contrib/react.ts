/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
