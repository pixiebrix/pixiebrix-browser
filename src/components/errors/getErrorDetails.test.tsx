/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from "@testing-library/react";
import { waitForEffect } from "@/testUtils/testHelpers";
import getErrorDetails from "./getErrorDetails";
import { type ErrorObject } from "serialize-error";

test("Template render error", () => {
  const error = {
    name: "Template render error",
    message:
      "(unknown path) [Line 1, Column 3]\n  parseAggregate: expected colon after dict key",
    stack:
      "Template render error: (unknown path) [Line 1, Column 3]\n  parseAggregate: expected colon after dict key\n    at Object._prettifyError (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/bundles/nunjucks.bundle.js:121:11)\n    at Template.render (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/bundles/nunjucks.bundle.js:3489:21)\n    at Environment.renderString (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/bundles/nunjucks.bundle.js:3331:17)\n    at Object.renderString (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/bundles/nunjucks.bundle.js:5652:14)\n    at default (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:56066:33)\n    at renderExplicit (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:55472:16)\n    at async Promise.all (index 1)\n    at async asyncMapValues (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:58604:20)\n    at async renderExplicit (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:55484:33)\n    at async Promise.all (index 0)",
  };

  const { title, detailsElement } = getErrorDetails(error);
  expect(title).toBe("Error");
  expect(render(detailsElement).asFragment()).toMatchSnapshot();
});

test("Input validation error", () => {
  const error = {
    name: "InputValidationError",
    schema: {
      $schema: "https://json-schema.org/draft/2019-09/schema#",
      type: "object",
      properties: {
        message: {
          type: ["string", "number", "boolean"],
          description: "The text/value you want to display in the alert",
        },
        type: {
          type: "string",
          description:
            "The alert type/style. 'window' uses the browser's native window alert dialog, which the user must dismiss",
          enum: ["window", "info", "success", "error"],
          default: "info",
        },
        duration: {
          type: "number",
          description:
            "Duration to show the alert, in milliseconds. Ignored for 'window' alerts",
          default: 2500,
        },
      },
      required: ["message"],
    },
    input: {
      type: "info",
      duration: 2500,
      message: {
        "@input": {
          description:
            "Habr is the largest resource for IT professionals in Europe. People visit Habr to discuss industry news and share own experience.",
          icon: "https://assets.habr.com/habr-web/img/favicons/apple-touch-icon-76.png",
          image: "https://habr.com/img/habr_en.png",
          title: "All posts in a row / Habr",
          language: "en",
          type: "website",
          url: "https://habr.com/en/all/",
          provider: "Habr",
        },
        "@options": {},
      },
    },
    errors: [
      {
        instanceLocation: "#",
        keyword: "properties",
        keywordLocation: "#/properties",
        error: 'Property "message" does not match schema.',
      },
      {
        instanceLocation: "#/message",
        keyword: "type",
        keywordLocation: "#/properties/message/type",
        error:
          'Instance type "object" is invalid. Expected "string", "number", "boolean".',
      },
    ],
    message: "Invalid inputs for block",
    stack:
      "InputValidationError: Invalid inputs for block\n    at throwIfInvalidInput (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:56223:15)\n    at async runBlock (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:55809:9)\n    at async blockReducer (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:55879:20)\n    at async reducePipeline (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:55989:26)\n    at async HTMLAnchorElement.<anonymous> (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/contentScript.js:51535:17)",
  };

  const { title, detailsElement } = getErrorDetails(error);
  expect(title).toBe("Invalid inputs for block");
  expect(render(detailsElement).asFragment()).toMatchSnapshot();
});

test("Network error", async () => {
  const error: ErrorObject = {
    cause: {
      config: {
        transitional: {
          silentJSONParsing: true,
          forcedJSONParsing: true,
          clarifyTimeoutError: false,
        },
        transformRequest: [],
        transformResponse: [],
        timeout: 0,
        xsrfCookieName: "XSRF-TOKEN",
        xsrfHeaderName: "X-XSRF-TOKEN",
        maxContentLength: -1,
        maxBodyLength: -1,
        headers: {
          Accept: "application/json, text/plain, */*",
        },
        url: "https://fail.com",
        method: "get",
      },
      request: {},
      isAxiosError: true,
      name: "Error",
      message: "Network Error",
      stack:
        "Error: Network Error\n    at createError (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:11801:15)\n    at _callee$ (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:10915:126)\n    at tryCatch (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:34316:40)\n    at Generator.invoke [as _invoke] (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:34547:22)\n    at Generator.throw (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:34372:21)\n    at asyncGeneratorStep (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:10798:24)\n    at _throw (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:10824:9)",
    },
    name: "ClientNetworkError",
    message: "Network error. No response received.",
    stack:
      "ClientNetworkError: No response received. Your browser may have blocked the request. See https://docs.pixiebrix.com/network-errors for troubleshooting information\n    at enrichBusinessRequestError (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:54834:12)\n    at async serializableAxiosRequest (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:43429:15)\n    at async handleMessage (chrome-extension://mpjjildhmpddojocokjkgmlkkkfjnepo/background.js:67712:22)",
  };

  const { title, detailsElement } = getErrorDetails(error);
  expect(title).toBe("Network error. No response received.");
  const { asFragment } = render(detailsElement);
  await waitForEffect();
  expect(asFragment()).toMatchSnapshot();
});

test("RemoteApiErrorDetail", async () => {
  const error: ErrorObject = {
    response: {
      data: { error: { message: "Invalid data" } },
      status: 400,
      statusText: "Bad Request",
    },
    name: "ProxiedRemoteServiceError",
    message: "Invalid data message.",
    stack: "ProxiedRemoteServiceError: Invalid data message",
  };

  const { title, detailsElement } = getErrorDetails(error);
  expect(title).toBe("Invalid data message.");
  const { asFragment } = render(detailsElement);
  await waitForEffect();
  expect(asFragment()).toMatchSnapshot();
});
