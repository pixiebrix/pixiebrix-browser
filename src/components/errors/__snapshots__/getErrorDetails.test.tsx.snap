// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input validation error 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <span>
        Errors
      </span>
      <ul>
        <li>
          #/properties: Property "message" does not match schema.
        </li>
        <li>
          #/properties/message/type: Instance type "object" is invalid. Expected "string", "number", "boolean".
        </li>
      </ul>
    </div>
    <div
      class="col"
    >
      <span>
        Rendered Args
      </span>
      <div
        class="root"
      >
        <ul
          style="border: 0px; padding: 0px; margin: 0.5em 0px 0.5em 0.125em; list-style: none; background-color: rgb(253, 255, 255);"
        >
          <li
            style="padding: 0px; margin: 0px;"
          >
            <ul
              style="padding: 0px; margin: 0px; list-style: none;"
            >
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    type:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "info"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    duration:
                  </span>
                </label>
                <span
                  style="color: rgb(243, 100, 51);"
                >
                  2500
                </span>
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px;"
              >
                <div
                  style="display: inline-block; padding-right: 0.5em; padding-left: 0px; cursor: pointer;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; color: rgb(66, 122, 245);"
                  >
                    ▶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <span>
                    message:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      {}
                    </span>
                     2 keys
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="col"
    >
      <span>
        Schema
      </span>
      <div
        class="root"
      >
        <ul
          style="border: 0px; padding: 0px; margin: 0.5em 0px 0.5em 0.125em; list-style: none; background-color: rgb(253, 255, 255);"
        >
          <li
            style="padding: 0px; margin: 0px;"
          >
            <ul
              style="padding: 0px; margin: 0px; list-style: none;"
            >
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    $schema:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "https://json-schema.org/draft/2019-09/schema#"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    type:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "object"
                </span>
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px;"
              >
                <div
                  style="display: inline-block; padding-right: 0.5em; padding-left: 0px; cursor: pointer;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; color: rgb(66, 122, 245);"
                  >
                    ▶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <span>
                    properties:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      {}
                    </span>
                     3 keys
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px;"
              >
                <div
                  style="display: inline-block; padding-right: 0.5em; padding-left: 0px; cursor: pointer;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; color: rgb(66, 122, 245);"
                  >
                    ▶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <span>
                    required:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      []
                    </span>
                     1 item
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Network error 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <span>
        Response
      </span>
      <div
        class="text-warning"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-exclamation-triangle fa-w-18 "
          data-icon="exclamation-triangle"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
            fill="currentColor"
          />
        </svg>
         PixieBrix does not have permission to access https://fail.com
      </div>
      <div>
        <div>
          PixieBrix did not receive a response. Possible causes:
        </div>
        <ul>
          <li>
            Your browser or another extension blocked the request. Check that PixieBrix has permission to the access the host. If PixieBrix is not showing a Grant Permissions button, ensure that the integration has an 
            <code
              class="px-0 mx-0"
            >
              isAvailable
            </code>
             section defined
          </li>
          <li>
            The remote server did not respond. Try the request again
          </li>
        </ul>
      </div>
    </div>
    <div
      class="col"
    >
      <span>
        Request Config
      </span>
      <div
        class="root"
      >
        <ul
          style="border: 0px; padding: 0px; margin: 0.5em 0px 0.5em 0.125em; list-style: none; background-color: rgb(253, 255, 255);"
        >
          <li
            style="padding: 0px; margin: 0px;"
          >
            <ul
              style="padding: 0px; margin: 0px; list-style: none;"
            >
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px;"
              >
                <div
                  style="display: inline-block; padding-right: 0.5em; padding-left: 0px; cursor: pointer;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; color: rgb(66, 122, 245);"
                  >
                    ▶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <span>
                    transitional:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      {}
                    </span>
                     3 keys
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 1.125em;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: default;"
                >
                  <span>
                    transformRequest:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      []
                    </span>
                     0 items
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 1.125em;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: default;"
                >
                  <span>
                    transformResponse:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      []
                    </span>
                     0 items
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    timeout:
                  </span>
                </label>
                <span
                  style="color: rgb(243, 100, 51);"
                >
                  0
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    xsrfCookieName:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "XSRF-TOKEN"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    xsrfHeaderName:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "X-XSRF-TOKEN"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    maxContentLength:
                  </span>
                </label>
                <span
                  style="color: rgb(243, 100, 51);"
                >
                  -1
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    maxBodyLength:
                  </span>
                </label>
                <span
                  style="color: rgb(243, 100, 51);"
                >
                  -1
                </span>
              </li>
              <li
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px;"
              >
                <div
                  style="display: inline-block; padding-right: 0.5em; padding-left: 0px; cursor: pointer;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; color: rgb(66, 122, 245);"
                  >
                    ▶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <span>
                    headers:
                  </span>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      {}
                    </span>
                     1 key
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block;"
                />
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    url:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "https://fail.com"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    method:
                  </span>
                </label>
                <span
                  style="color: rgb(64, 175, 108);"
                >
                  "get"
                </span>
              </li>
              <li
                style="padding-top: 0.25em; padding-right: 0px; margin-left: 0.875em; word-wrap: break-word; padding-left: 1.25em; text-indent: -0.5em; word-break: break-all; white-space: pre-wrap;"
              >
                <label
                  style="display: inline-block; color: rgb(66, 122, 245); margin: 0px 0.5em 0px 0px;"
                >
                  <span>
                    data:
                  </span>
                </label>
                <span
                  style="color: rgb(241, 89, 80);"
                >
                  undefined
                </span>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Template render error 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      (unknown path) [Line 1, Column 3]
  parseAggregate: expected colon after dict key
    </div>
  </div>
</DocumentFragment>
`;
